### ISSUE TITLE
Improve shallow rendering for use with higher-order components

### ISSUE THREAD

**Issue by lencioni** [airbnb/enzyme#539]
When using shallow rendering with a component that is wrapped by a higher order component, you need to take some extra steps to make shallow rendering give you the structure you expect. One way is to shallow render the wrapped component, find the unwrapped component within this structure, and then shallow render that. e.g.

```jsx
const wrapper = shallow(<Foo />).find('Foo').shallow();
```

Of course, you can also export the unwrapped component as a named export, and use that to find by reference as well.

```jsx
import Foo, { Foo as UnwrappedFoo } from './Foo';

...

const wrapper = shallow(<Foo />).find(UnwrappedFoo).shallow();
```

This bums me out a little, because if you have a bunch of tests for a component, and then later decide to wrap that component in a HOC, you need to go back and update all of your tests in this way. Using the string approach feels less than ideal, and the reference approach feels a bit cumbersome when applied on a large scale. I'm wondering if we can come up with a solution that will work well for this scenario. Here are some thoughts.

- In #250, @lelandrichardson proposed "Mixed-depth shallow rendering mode", which would allow you to pass an `expand` option to `shallow` with references to components to "expand" the shallow render tree into. I think this would work if your component is only wrapped by a single HOC, but if you wrap it in multiple HOCs, you end up with the same problem again, because you have multiple levels that you need to expand, but only a reference to the top level.

- We could add an `until` option to shallow that would take a reference to a component or a string reference. This would reduce some of the boilerplate of `.find('Foo').shallow()` just a little, but I think it has many of the same disadvantages that we currently have.

- Another possibility is to provide a `depth` option to `shallow`, that would take a number of components to drill down into. I think that this would be more comfortable than what we currently do, but it still doesn't fully solve the problem--you still have to know to add this option to every place you want to use shallow rendering with wrapped components. And, if you wrap a component in another HOC, you need to increment all of the values.

- We could add a static property on HOCs that signify that they are HOCs. Enzyme could then look for this static property when shallow rendering, and dig deeper. And, if you find yourself in a situation where you want to avoid this behavior, we could add an option to `shallow` to opt-out, but maybe that's YAGNI. The downside is that this would only work for HOCs that opt in to this behavior, so it likely wouldn't solve the problem when using third-party HOCs--unless it somehow caught on and spread to all corners.

- What if we allowed people to register a list of regexes used to expand components when shallow rendering, matching on their names. Using this, folks could configure their testing environment to do what they want, and then not have to think about it much any more. At Airbnb, [we use a naming convention of `functionName(ComponentName)` for HOCs](https://github.com/airbnb/javascript/tree/master/react#naming) which we could configure all be automatically expanded with a regex like `/\([^\)]+\)/`. I like this option because I think it would solve this issue for us with little configuration, but I dislike it because it makes shallow rendering somewhat magical and might cause unexpected results.

Do any of these sound good? Should we employ multiple strategies? Are there other possibilities?

**Comment by nfcampos** [comment]
You can also do 
```js
const wrapper = shallow(<Foo />).first().shallow();
```
I definitely feel this pain, but I'm not sure what the right solution is. But I do lean more towards defining the `depth`

**Comment by lencioni** [comment]
That's true, but I think you have to tag on another `.first().shallow()` for each HOC.

**Comment by nfcampos** [comment]
yes if you have 3 HOCs it becomes the rather terrible
```js
const wrapper = shallow(<Foo />).first().shallow().first().shallow().first().shallow();
```

**Comment by ljharb** [comment]
The only option I like there is #250. Even with multiple HOCs, I think that would work transparently. You'd still always need to export the pure version, but I think that's just an inherent limitation in HOCs that enzyme can't gracefully overcome.

**Comment by lencioni** [comment]
If I understand that proposal correctly, you would need to export each level of HOC that you wrap your component in, and then pass those all as references to the `expand` option. I think I would rather use `.find('Foo').shallow()`.

**Comment by Aweary** [comment]
Ideally, I think it would be nice to have an API would be able to find a component at any level in the potential render tree for a component, even if it was too deep to be rendered with a `shallow` render. 

```js
var deepWrapper = wrapper.shallowFromDeepChild('SomeDeepComponent');
```

Though I think that would be difficult to implement. I like #250 but if you have to export every level of HOC then it doesn't seem to bring any advantage over just testing the exported component directly. It's definitely a hard problem.

**Comment by lencioni** [comment]
@Aweary I think your suggestion is similar to my thought about adding an `until` option, which would keep going until it rendered something that matched the selector.

**Comment by CurtisHumphrey** [comment]
I suggest instead of a list to expand have a list of do not expand. For me and our tests, most of the time we use shallow because we have components inside of it we do not want to expand (big things maybe directly connect to redux etc).  We know this list as build the widget and tests. Then if we later wrap it with HOC that list doesn't change and nothing in the tests have to change, problem solved.

**Comment by ljharb** [comment]
Ideally we'd want both options I think: "only expand through these" or "expand unless it's one of these".

**Comment by ljharb** [comment]
re https://github.com/airbnb/enzyme/issues/539#issuecomment-239843015 - specifically, as a separate API from `shallow` - and I don't think we should allow any sort of "depth" number, since that tightly couples tests to JSX structure.

**Comment by CurtisHumphrey** [comment]
I second @ljharb on both points (both options) and no depth number

**Comment by oliviertassinari** [comment]
For anyone looking for a temporary solution, we are using a `until` operator at @doctolib as we use quite some HoC:

cc @matthieuprat for the credit.
<details>
  <summary>enzyme/operator/until.js</summary>
```jsx
export default function until(selector, {context} = this.options) {
  if (this.isEmptyRender() || typeof this.node.type === 'string')
    return this;

  const instance = this.instance();
  if (instance.getChildContext) {
    context = {
      ...context,
      ...instance.getChildContext(),
    };
  }

  return this.is(selector)
    ? this.shallow({context})
    : this.shallow({context})::until(selector, {context});
}
```
</details>

That we use like this:
```jsx
const AppBanner = ({banner, children}) => {
  return (
    <Banner>
      {children}
    </Banner>
  );
};

export default compose(
  pure,
  connect(({
    globalMessagesBanner$,
  }) => ({
    banner: globalMessagesBanner$,
  })),
)(AppBanner);
```

```jsx
assert.strictEqual(
  wrapper.until('[banner]').find(Banner).props().children,
  'key: common.error'
);
```

**Comment by matthieuprat** [comment]
Here is a Gist with a bunch of tests for this `until` operator to make it more "graspable": https://gist.github.com/matthieuprat/5fd37abbd4a4002e6cfe0c73ae54cda8

**Comment by elodszopos** [comment]
Here's a little helper that I use. Works for multiple nesting too (multiple decorators / HoC wrapping).

```
export function renderComponent(Component, defaultProps = {}) {
  return function innerRender(props = {}, nestLevel) {
    let rendered = shallow(<Component {...defaultProps} {...props} />);

    if (nestLevel) {
      let repeat = nestLevel;

      while (repeat--) {
        rendered = rendered.first().shallow();
      }
    }

    return rendered;
  };
}
```

And then:

```
const renderer = renderComponent(Foo); // defined in describe
const wrapper = renderer({ someProp: someValue }, 1); // @connected Foo, to be used in an it block
```

It's a temporary solution really. I'd like to fully support this effort and if there's anything I can help with, I will. 

I'm also all for what @ljharb said in [his comment above](https://github.com/airbnb/enzyme/issues/539#issuecomment-239843015). The temporary solution that I have tightly couples to JSX and is volatile. Far from ideal.

**Comment by kumar303** [comment]
This is a helper that is working for me so far. It's nice because you can wrap as many HOCs around your component as you need to without having to update any test code. It's pretty similar to @oliviertassinari's approach but takes a component class instead of a selector.

```js
import { shallow } from 'enzyme';

export function shallowUntilTarget(componentInstance, TargetComponent, {
  maxTries = 10,
  shallowOptions,
} = {}) {
  let root = shallow(componentInstance, shallowOptions);

  if (typeof root.type() === 'string') {
    // If type() is a string then it's a DOM Node.
    // If it were wrapped, it would be a React component.
    throw new Error('Cannot unwrap this component because it is not wrapped');
  }

  for (let tries = 1; tries <= maxTries; tries++) {
    if (root.is(TargetComponent)) {
      // Now that we found the target component, render it.
      return root.shallow(shallowOptions);
    }
    // Unwrap the next component in the hierarchy.
    root = root.first().shallow(shallowOptions);
  }

  throw new Error(
    `Could not find ${TargetComponent} in React instance: ${componentInstance}; 
    gave up after ${maxTries} tries`
  );
}
```

Here is an example of how I use it. In `SomeComponent.js`:

```js
import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';

export function SomeComponentBase() {
  return <div>This is the unwrapped component that we want to test</div>;
}

// Pretend that this is a typical Redux mapper.
const mapStateToProps = (state) => ({});

export default compose(
  connect(mapStateToProps),
)(SomeComponentBase);
```

Example of rendering it for a test:

```js
import SomeComponent, { SomeComponentBase } from './SomeComponent';
import { shallowUntilTarget } from './utils';

// This unwraps SomeComponent until it finds SomeComponentBase:
const root = shallowUntilTarget(<SomeComponent />, SomeComponentBase);
```

**Comment by ljharb** [comment]
Linking to #250.

**Comment by goldenshun** [comment]
I wasn't able to get the "shallow until render" approaches to work, so went with an approach that mocks the specific component I want to avoid rendering.  Jest mocks are a little quirky, but they did the job.  Especially helpful when testing Apollo's Query/graphql components.

https://medium.com/@sconnolly17/testing-graphql-container-components-with-react-apollo-and-jest-9706a00b4aeb

**Comment by kumar303** [comment]
> Honestly you've changed your design, your Component Under test is now doing more.

Yes but the new HOC is just an implementation detail. The interface of your component did not change so the unit tests for it should not have to change.

**Comment by dschinkel** [comment]
Yea I just don't think it's a big deal in those cases honestly personally.

**Comment by dschinkel** [comment]
>We could add an until option to shallow that would take a reference to a component or a string reference. This would reduce some of the boilerplate of .find('Foo').shallow() just a little, but I think it has many of the same disadvantages that we currently have

@lencioni can you site what you mean by:
>I think it has many of the same disadvantages that we currently have

**Comment by lencioni** [comment]
Oh boy I wrote that a long time ago and before we added dive. I don't remember exactly what I meant, but maybe I was thinking about the possibility that Enzyme could be configured to do this automatically so when you introduce or remove an HOC, you don't need to change a bunch of tests. Although, that certainly has its own set of tradeoffs and that probably would be worse in a number of other ways.

**Comment by dalvallana** [comment]
I'm into something similar here: `MyComponent` calls a class method under certain conditions in the constructor, so I want to stub this method _before_ `shallow`ing `MyComponent`, in order to test it is called when it should. Normally I would do something like:
``` 
sinon.stub(MyComponent.prototype, 'myMethod');
```
But when `MyComponent` is wrapped in a HOC, I end up doing:
```
sinon.stub(MyComponent.prototype.constructor.WrappedComponent.prototype, 'myMethod');
```
which is frankly awful...

I honestly don't know if enzyme provides a nicer way of doing this... Maybe I'm missing something.

**Comment by ljharb** [comment]
@dalvallana you either have to do that, or to export the wrapped component directly - neither of which is nice. In that specific case, however, you should be able to shallow render the HOC, and *before* calling `.dive()`, do `spyOn(wrapper.childAt(0).type().prototype, 'myMethod')` (i believe that will work).

**Comment by slikts** [comment]
I wish this was solved in Enzyme instead of there being a potpourri of snippets, gists and unmaintained one-off packages to choose from. Not having official support suggests that there's something unsound about diving based on a selector, but this thread doesn't really explicate what the issues would be.

**Comment by ljharb** [comment]
@slikts typically you'd use one `.dive()` per HOC, and not require any selectors.

**Comment by slikts** [comment]
It's not required in the sense that there is a workaround (chaining `.dive()`), but the depth of nesting is an implementation detail. For example, I might refactor `withABC()` into `withA()`, `withB()`, etc., and the test failing would be a false negative.

**Comment by goldenshun** [comment]
@sliks I think the root of the problem is that Enzyme provides too many utilities that are dependent on the structure of the component tree rather than something closer to what the user is seeing. This is the reason why I am moving away from Enzyme and towards `react-testing-library`. With RTL you would solve this problem with jest mocks, similar to what I wrote about in this post:

https://blog.usejournal.com/testing-graphql-container-components-with-react-apollo-and-jest-9706a00b4aeb

**Comment by ljharb** [comment]
@slikts yes, that is true, but your tests for the component *are* testing some of the implementation details, and that's ok.

@goldensunliu good testing is a balance of surface testing (like what RTL and `mount` provides) and unit testing (like what `shallow` provides). Only using one technique is always a mistake. Separately, using mocks and stubs makes tests more brittle, and should be avoided.

**Comment by goldensunliu** [comment]
@ljharb I think you meant to tag @goldenshun. no worries.

**Comment by goldenshun** [comment]
@ljharb <https://github.com/ljharb> The entire premise of this thread is
that using shallow with higher order components is brittle. Mocks can be
brittle too but in this case they are more stable than shallow.

**Comment by ljharb** [comment]
@goldenshun i suppose that’s subjective; i don’t find that to be the case.

Certainly there is an improvement to be made here, but the brittleness you’re referring to is “you have to match the count of HOCs with the count of dives”. In practice, this is only mildly annoying - the brittleness of mocks is that your code will fail in production but your tests won’t show that, because the mocks aren’t up to date and mask the errors.

**Comment by goldenshun** [comment]
@ljharb Your criticism of mocks is true for both jest mocks and shallow
rendering. They are both forms of mocking because they both halt rendering
of the component tree at a certain later of the component tree. As with any
unit test, they will not detect all classes of error.

My submission is that jest mocks are a more reliable mechanism for the case
at hand because they will be more resilient to changes in the component
structure.

**Comment by ljharb** [comment]
In practice, I’ve found the opposite to be true - and while changing the component structure shouldn’t change tests for other things, the tests for the component being changed shouldn’t necessarily be completely resilient to changes in structure.

**Comment by dschinkel** [comment]
This thread drives me nuts.  Has everyone just ignored what I wrote above?  it's YOUR DESIGN not ENZYME.

>I might refactor withABC() into withA(), withB(), etc., and the test failing would be a false negative
Then your damn method is doing TOO MUCH in the first place.  If you were To TDD you'd see that earl and already have broken it out and realized this.  Test after, you're fucked.

>@slikts I think the root of the problem is that Enzyme provides too many utilities that are dependent on the structure of the component tree rather than something closer to what the user is seeing. This is the reason why I am moving away from Enzyme and towards react-testing-library

Don't start with this bullshit.  I'm so tired of hearing about Dodd's garbage.  Your tests should first test small components under your UI.  Your UI is not the behavior you're testing, it's the USE CASE.  Write smaller tests first. Then go back and add the stupid integration tests.

> Test resiliency

This is so bad.  Dodds and others think tests should never break.  They're missing the fact that if your design changes enough, your tests WILL break. Don't look for hacks to get around that. That's jut a fact. This goes back again to writing smaller tests first, breaking stuff out to smaller things early in your production code.  Don't look for bandaids like nested dives when this is telling you that your design is bad.

Test resiliancy is not what you should be striving for. You should be striving for TESTABLE CODE.  You should be striving for SMALL FUNCTIONS, SMALL COMPONENTS.  You should be striving for simple design, decoupling things early, etc.

Again, stop writing tests after, and you'll see the light.  Honestly, I get so sick of this.

**Comment by dschinkel** [comment]
@goldensunliu this is exactly why I hate Apollo, tons of nested HOCs.  It's bad design period.

**Comment by hmmChase** [comment]
I have a class component called `IdeaCard.js` that using `styled-components` and `react-apollo` that looks like:

```
render() {
    return (
      <sc.li>
        <Mutation
          mutation={query.DELETE_IDEA_MUTATION}
          variables={{ id: this.props.id }}
        >
          {deleteIdea => (
            <sc.deleteBtn
              type="button"
              onClick={e => this.handleClickDeleteBtn(e, deleteIdea)}
            />
          )}
        </Mutation>

        <Mutation
          mutation={query.UPDATE_IDEA_MUTATION}
          variables={{ id: this.props.id, content: this.state.content }}
        >
          {updateIdea => (
            <sc.ideaInput
              type="text"
              value={this.state.content}
              onChange={e => this.handleChangeideaInput(e, updateIdea)}
            />
          )}
        </Mutation>
      </sc.li>
    );
  }
```

If I do:
```
wrapper = shallow(
    <MockedProvider mocks={mockQueries} addTypename={false}>
        <IdeaCard {...mockProps} />
    </MockedProvider>
);
```

And then:
```
console.log(wrapper.dive().dive().dive().debug());
```

I get:
```
<IdeaCardstyle__li>
    <Mutation mutation={{...}} variables={{...}}>
        [function]
    </Mutation>
    <Mutation mutation={{...}} variables={{...}}>
        [function]
    </Mutation>
</IdeaCardstyle__li>
```

Is there something I can do test this, or do I need to use `mount` instead of `shallow`?

Thanks for any help.

**Comment by ljharb** [comment]
@hmmChase use the new `wrappingComponent` option to pass `MockedProvider`, which will save you a dive; but otherwise, what do you want to test? From that debug, you can `.find` `Mutation` and use `.renderProp` to test what the function returns, you can assert on the props passed to Mutation (and delegate that testing responsibility to Mutation's tests)… please file a new issue if you still have questions.

**Comment by sktguha** [comment]
Hi, any update on this ? are there plans to implement something like shallowuntil or like list of nodes which are to be expanded or not expanded,   as discussed ?
